nodewatcher/nodewatcher.py

```diff
@@ -183,57 +260,94 @@ def main():
     else:
         data = {_CURRENT_IDLETIME: 0}
 
+    instance_id = get_instance_id()
+    hostname = get_hostname()
+
+    dynamodb_resource = boto3.resource('dynamodb', region_name=region, config=proxy_config)
+    try:
+        dynamodb_table = dynamodb_resource.Table(table_name)
+    except ClientError as e:
+        log.critical('Creating a table client failed with exception %s. Exiting' % e)
+        raise
+
     stack_creation_complete = False
     termination_in_progress = False
     while True:
-        # if this node is terminating sleep for a long time and wait for termination
+        # If termination is in progress, keep the daemon running as a no-op
         if termination_in_progress:
+            log.info("terminating %s" % instance_id)
             time.sleep(300)
             log.info('%s is still terminating' % hostname)
             continue
         time.sleep(60)
         if not stack_creation_complete:
-            stack_creation_complete = stackCreationComplete(stack_name, region, proxy_config)
+            stack_creation_complete = get_stack_creation_status(stack_name, region, proxy_config)
             log.info('%s creation complete: %s' % (stack_name, stack_creation_complete))
             continue
         asg_conn = boto3.client('autoscaling', region_name=region, config=proxy_config)
 
-        has_jobs = hasJobs(s, hostname)
+        has_jobs = node_has_jobs(scheduler, hostname)
         if has_jobs:
             log.info('Instance has active jobs.')
             data[_CURRENT_IDLETIME] = 0
         else:
-            if maintainSize(asg_name, asg_conn):
+            min_size, current_size = get_desired_sizes(scheduler, asg_name, asg_conn)
+            if current_size <= min_size:
```

Comment:
We should consider resetting CURRENT_IDLETIME in cases when the node is not marked for termination because the min_size of the cluster is reached or when there are pending jobs that are waiting to be submitted. Currently CURRENT_IDLETIME is kept at the previously computed value, meaning the idle time might not be guaranteed under certain circumstances (e.g. the CURRENT_IDLETIME was already bigger than threshold but node is not terminated because min_size is reached. In the following cycles this node will be terminated right away when cluster size grows)